### PR TITLE
[MRG] Replacing `ValueError` in `sourmash_tst_utils.runscript(...)` with a custom exception

### LINKED ISSUE #1611: replace `ValueError` in `sourmash_tst_utils.runscript(...)` with a custom exception
right now, failing command line tests raise a `ValueError`, which is confusing. Specifically,

```
def test_run_sourmash(runtmp):
   ...
   runtmp.sourmash(*args)
```
will raise a `ValueError` if the exit status of the `sourmash` script is non-zero, indicating failure (e.g. `sys.exit(-1)`, or an uncaught exception).

This behavior is defined in `tests/sourmash_tst_utils.py`, code:
```
        if self.last_result.status:
            raise ValueError(self)
```

There is no particular reason for this, it's just an arbitrary decision I made several years ago.

After some discussion, @luizirber and @bluegenes and I agree that this should be changed to a custom exception, perhaps `sourmash_tst_utils.SourmashCommandFailed`.

To resolve this issue,
* define a new exception class, `SourmashCommandFailed`, in `sourmash_tst_utils.py`; it should inherit from `Exception`;
* change the behavior in `sourmash_tst_utils.py`
* fix all of the tests that use `with pytest.raises(ValueError)` to instead use `with pytest.raises(SourmashCommandFailed)`

### PR DESCRIPTION
**Pull request by keyabarve**
Fixes #1611 

TODO: 
- define a new exception class, `SourmashCommandFailed`, in `sourmash_tst_utils.py`; it should inherit from `Exception`;
- change the behavior in `sourmash_tst_utils.py`
- fix all of the tests that use `with pytest.raises(ValueError)` to instead use `with pytest.raises(SourmashCommandFailed)`

### CODE REVIEW

**Comment by keyabarve** [comment]
@ctb I have a couple of questions: 
- Should this class be defined in `sourmash_tst_utils.py`? 
- If so, what does this mean: "change the behavior in `sourmash_tst_utils.py`"?
- What does this mean: " it should inherit from `Exception`"?

**Comment by ctb** [comment]
> @ctb I have a couple of questions:
> 
>     * Should this class be defined in `sourmash_tst_utils.py`?

sure!

>     * If so, what does this mean: "change the behavior in `sourmash_tst_utils.py`"?

The new exception should be used in the code to signal that the script has exited with an error.

>     * What does this mean: " it should inherit from `Exception`"?

Exceptions are part of a class hierarchy in Python, so any new exception needs to have a base class. This affects how you catch and handle exceptions: If, for example, you defined a new exception as a subclass of `ValueError`, then `try: ... except ValueError: ...` would catch the new exception as well as any other `ValueError`.

In this case, I think the new exception should be a direct subclass of `Exception`, to reduce confusion about its connection with other exceptions used in soumash.

**Comment by keyabarve** [comment]
@ctb Could you provide some hints on how and where to get started?

**Comment by ctb** [comment]
hi @keyabarve I think your list of TODO items in the comment for this PR is a decent checklist. I've given you some example code over in https://github.com/sourmash-bio/sourmash/pull/1679.

**Comment by keyabarve** [comment]
@ctb I have added the `SourmashCommandFailed` class and just the `_init(..)` function to it. I have used `class SourmashError(Exception)` in `src/sourmash/exceptions.py` as a reference.

**Comment by keyabarve** [comment]
I have written the `SourmashCommandFailed` class in `tests/sourmash_tst_utils.py`. I will start changing all the tests where `ValueError` is raised.

**Comment by keyabarve** [comment]
In some of the tests, where I’ve replaced `with pytest.raises(ValueError) as exc:` with the command `with pytest.raises(utils.SourmashCommandFailed) as exc:` and the assert statement outside this block is in this format: `assert "..." in str(exc.value)`, this is an example of the error that I’m getting:
```
E       AssertionError: assert 'taxonomy table already exists' in ''
E        +  where '' = str(SourmashCommandFailed())
E        +    where SourmashCommandFailed() = <ExceptionInfo SourmashCommandFailed() tblen=2>.value
```
If I change the assert statement to: `assert "..." in c.last_result.err`, the tests pass. I was wondering if it would be okay to change the assert statement in this manner.

**Comment by ctb** [comment]
hi keya, ideally, you would need to only change the name of the exception, and nothing else.

In this case, it looks like you are not executing the parent class constructor properly and so the new exception class is not storing the message string. You definitely want the exception class to store the message string, in general!

To fix this, you could either call `Exception.__init__` with the full set of arguments passed to `SourmashCommandFailed.__init__`, or use `super` for single inheritance, as per [this article](https://realpython.com/python-super/.

Note also that you can do a `from sourmash_tst_utils import SourmashCommandFailed` to avoid the `utils.` prefix if you like.

**Comment by keyabarve** [comment]
> hi keya, ideally, you would need to only change the name of the exception, and nothing else.
> 
> In this case, it looks like you are not executing the parent class constructor properly and so the new exception class is not storing the message string. You definitely want the exception class to store the message string, in general!
> 
> To fix this, you could either call `Exception.__init__` with the full set of arguments passed to `SourmashCommandFailed.__init__`, or use `super` for single inheritance, as per [this article](https://realpython.com/python-super/.
> 
Okay, thanks I'll take a look at this.

> Note also that you can do a `from sourmash_tst_utils import SourmashCommandFailed` to avoid the `utils.` prefix if you like.
>
Sure, I'll do that.

**Comment by keyabarve** [comment]
@ctb I have made all the changes you requested and the tests now pass. This is ready for review.

File: tests/sourmash_tst_utils.py [hunk 1]
```diff
@@ -186,7 +192,8 @@ def run_sourmash(self, *args, **kwargs):
         self.last_result = runscript('sourmash', args, **kwargs)
 
         if self.last_result.status:
-            raise ValueError(self)
+            # raise ValueError(self)
```
Comment (ctb) [inline tests/sourmash_tst_utils.py]: please remove commented out line

**Review by ctb — CHANGES_REQUESTED** [review]
one minor change ;)

**Comment by ctb** [comment]
reminder - please change the title to MRG.

**Comment by keyabarve** [comment]
@ctb Please review.

**Comment by ctb** [comment]
hi keya, a test is failing because you merged from `latest` and a new test from that merge used the old `ValueError`. Could you please fix?

**Comment by keyabarve** [comment]
> hi keya, a test is failing because you merged from `latest` and a new test from that merge used the old `ValueError`. Could you please fix?
>
Alright!

**Comment by keyabarve** [comment]
@ctb Please review.

**Review by ctb — APPROVED** [review]

**Comment by ctb** [comment]
🎉

**Pull request merged**